### PR TITLE
[ADD] Python3 compatibility

### PR DESCRIPTION
**Pull request by StefanRijnhart**

### CODE REVIEW

**Comment by PieterPaulussen** [comment]
I tried these changes and I got a successful build after some trying. (Had some issues with Pillow)

**Comment by mileo** [comment]
Build sucessful with commands:

```
virtualenv . --python=python3
bin/pip install zc.buildout=2.8.0
bin/buildout
```
 But i had to delete my cache folder ~/.buildout/...

**Comment by mileo** [comment]
I'm doing the same test with python2 and Odoo 10.0. I will post the feedback soon.

**Comment by mileo** [comment]
LGTM: TESTED in 10.0 and 11.0

**Comment by mstuttgart** [comment]
When I try install pytz, it raise this erros:

```bash
  File "/opt/odoo-11/src/anybox.recipe.odoo/anybox/recipe/odoo/base.py", line 678, in read_odoo_setup
    'setup.py: ' + exception.message)
AttributeError: 'NameError' object has no attribute 'message'
```
`message` atributte was removed from Python 3, I guess. Muts be changed to:

```python
str(exception)
```
I find this lines that use `exception.message`:
* https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/eeceeb9a218d8bd45dc16ab91e08cb4d88ac27b5/anybox/recipe/odoo/base.py#L579
* https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/eeceeb9a218d8bd45dc16ab91e08cb4d88ac27b5/anybox/recipe/odoo/base.py#L655
* https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/eeceeb9a218d8bd45dc16ab91e08cb4d88ac27b5/anybox/recipe/odoo/base.py#L665
* https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/eeceeb9a218d8bd45dc16ab91e08cb4d88ac27b5/anybox/recipe/odoo/base.py#L667
* https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/eeceeb9a218d8bd45dc16ab91e08cb4d88ac27b5/anybox/recipe/odoo/base.py#L678

**Comment by StefanRijnhart** [comment]
Thanks @mstuttgart for reporting this problem! Can you test with the new commit?

**Comment by mstuttgart** [comment]
Yes :+1: I will post the feedback here.

**Comment by mstuttgart** [comment]
@StefanRijnhart its work :+1:

**Comment by PierreFaniel** [comment]
Works like a charm ! Thanks for the good work !

**Comment by PierreFaniel** [comment]
Hello Stefan, 

I encountered an issue with the .keys() function which in P3 doesn't return a list as in P2.

So you might want to fix the following line By `for k in list(sys.modules.keys()):` which both works in P2 and P3:
https://github.com/StefanRijnhart/anybox.recipe.odoo/blob/3c67e9c00b7dae1cc34beda1e079e0ee80217ec4/anybox/recipe/odoo/base.py#L415

Again, cheers for the migration !

**Comment by StefanRijnhart** [comment]
@PierreFaniel thanks for noting. I see that the iterator breaks because the dictionary is modified during the iteration. Your solution should work find. I added it in https://github.com/anybox/anybox.recipe.odoo/pull/108/commits/6e74105a73aa429b266fd4c55f6d206f9d5d7a27

File: anybox/recipe/odoo/base.py [hunk 1]
```diff
@@ -25,19 +28,30 @@
...
-from urlparse import urlparse
+try:
+    import httplib  # Python 2
+except ImportError:
+    from http import client as httplib  # Python 3
+from email import utils as email_utils
+try:
+    from urlparse import urlparse  # Python 2
+except ImportError:
+    from urllib import parse as urlparse  # Python 3
```
[the remark this answers is not in the document]
Comment (StefanRijnhart) [reply anybox/recipe/odoo/base.py]: Thanks @sebalix! Fixed in https://github.com/anybox/anybox.recipe.odoo/pull/108/commits/6f8c3c5608f003dda5ae99206e48695172c6135f

**Comment by StefanRijnhart** [comment]
@petrus-v Would you agree to merging this? This branch has had lots of testing and additional fixes from other contributors. While there may be remaining problems that are not uncovered because the tests are currently broken on the main branch, I think we should just go ahead with this one.

File: anybox/recipe/odoo/runtime/patch_odoo.py [hunk 2]
```diff
@@ -12,10 +12,9 @@ def do_patch(gevent_script_path):
     isolated from the actual process management logic in the original.
     """
 
-    try:
-        from odoo.service.server import PreforkServer, stripped_sys_argv
-        from odoo.release import version_info
-    except ImportError:
+    from odoo.service.server import PreforkServer, stripped_sys_argv
+    from odoo.release import version_info
+    if False:
```
Comment (sebalix) [inline anybox/recipe/odoo/runtime/patch_odoo.py]: This import does not support Odoo < 10.0 (just encounter the problem on 9.0). You think we should merge anyway?

**Comment by ventilooo** [comment]
@StefanRijnhart @sebalix is that PR still plan to be merge ?

**Comment by sebalix** [comment]
I tested this branch on different projects: 8.0, 9.0, 10.0 and 11.0 with success.

@StefanRijnhart Can I merge?
Don't know what is the preferred way for this repo, create a merge commit or rebase and merge?

**Comment by StefanRijnhart** [comment]
I would prefer a merge so that the attribution of all contributors stays intact.

**Comment by StefanRijnhart** [comment]
Thank you @sebalix and others for your reviews, testing and contributions!

**Pull request merged**